File: src/lib/ipc/greet.ts
import { settings } from '$lib/config';
import { trace } from '@opentelemetry/api';
import { invoke } from '@tauri-apps/api';
import { IPCMessage } from './message';

/**
 * Call greet IPC.
 * @param name Name of user to greet.
 * @returns Greeting message.
 */
export async function greet(name: string): Promise<string> {
	const tracer = trace.getTracer(settings.APPLICATION_ID);
	const span = tracer.startSpan('ipc/greet');
	span.setAttribute('ipc.request.name', name);

	const response: IPCMessage<string> = await invoke('greet', { message: new IPCMessage(name) });
	const greeting = response.body;

	span.setAttribute('ipc.response.greeting', greeting);
	span.end();

	return greeting;
}

File: src-tauri/src/ipc/greet.rs
use opentelemetry::{global,
                    trace::{Span, Tracer},
                    KeyValue};

use crate::{config,
            ipc::{error::IPCError, message::IPCMessage},
            service};

/// Greet user.
#[tauri::command]
pub fn greet(message: IPCMessage<String>) -> Result<IPCMessage<String>, IPCError> {
    let cx = message.extract_cx();
    let _cx_guard = cx.attach();
    let tracer = global::tracer(config::APPLICATION_ID);
    let mut span = tracer.start("ipc/greet");

    let greeting = service::greet(message.body());
    span.set_attribute(KeyValue::new("ipc.response.greeting", greeting.clone()));

    Ok(IPCMessage::builder().body(greeting).build())
}

#[cfg(test)]
mod tests {
    use anyhow::Result;

    use super::{greet, IPCMessage};

    #[test]
    fn test_greet() -> Result<()> {
        let message = IPCMessage::builder().body("Libra".to_string()).build();
        let greeting = greet(message);

        assert_eq!(greeting.unwrap().body(), "Hello, Libra! (good)");

        Ok(())
    }
}
